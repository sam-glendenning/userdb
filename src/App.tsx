import React from "react";
import "./App.css";
import { useUserList } from "./requests";

const App = (): React.ReactElement => {
  const { data } = useUserList();

  return (
    <div className="App">
      <p>User Database!</p>
      {data && data.data.map((user) => <p>Email: {user.email}</p>)}
    </div>
  );
};

export default App;
