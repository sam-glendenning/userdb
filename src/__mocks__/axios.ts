import axios from "axios";

const requests = {
  get: jest.fn(() => Promise.resolve({ data: {} })),
  post: jest.fn(() => Promise.resolve({ data: {} })),
  delete: jest.fn(() => Promise.resolve({ data: {} })),
  CancelToken: axios.CancelToken,
};

export default requests;
