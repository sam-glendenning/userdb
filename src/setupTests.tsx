// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import "@testing-library/jest-dom";
import { render, type RenderOptions } from "@testing-library/react";
import React from "react";

const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
    logger: { log: console.log, warn: console.warn, error: jest.fn() },
  });

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  queryClient: QueryClient;
}

export const renderComponentWithQueryClient = (
  ui: React.ReactElement,
  { queryClient, ...renderOptions }: ExtendedRenderOptions = {
    queryClient: createTestQueryClient(),
  }
) => {
  const Wrapper = ({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element => (
    <QueryClientProvider client={queryClient ?? createTestQueryClient()}>
      {children}
    </QueryClientProvider>
  );

  return {
    queryClient,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
};
