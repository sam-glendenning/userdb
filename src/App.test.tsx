import React from "react";
import {
  RenderResult,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import App from "./App";
import { renderComponentWithQueryClient } from "./setupTests";
import usersJson from "./users.json";
import axios from "axios";

describe("App", () => {
  const createView = (): RenderResult =>
    renderComponentWithQueryClient(<App />);

  beforeEach(() => {
    (axios.get as jest.Mock).mockResolvedValue({
      data: usersJson,
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", async () => {
    const { asFragment } = createView();

    await waitForElementToBeRemoved(screen.queryByText("Loading users..."));

    expect(asFragment()).toMatchSnapshot();
  });
});
