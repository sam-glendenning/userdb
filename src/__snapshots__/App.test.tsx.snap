// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <p>
      User Database!
    </p>
    <br />
    <div>
      <p>
        Here are all the current users (this list from the API does not update):
      </p>
      <br />
      <p>
        1. michael.lawson@reqres.in
      </p>
      <p>
        2. lindsay.ferguson@reqres.in
      </p>
      <p>
        3. tobias.funke@reqres.in
      </p>
      <p>
        4. byron.fields@reqres.in
      </p>
      <p>
        5. george.edwards@reqres.in
      </p>
      <p>
        6. rachel.howell@reqres.in
      </p>
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Add User
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Remove User
      </button>
    </div>
    <br />
  </div>
</DocumentFragment>
`;
